File: bot/mirai/src/main/java/moe/dituon/petpet/bot/qq/mirai/GroupMessageEventHandler.java
package moe.dituon.petpet.bot.qq.mirai;

import lombok.extern.slf4j.Slf4j;
import moe.dituon.petpet.bot.qq.ReplyType;
import moe.dituon.petpet.bot.qq.avatar.QQAvatarRequester;
import moe.dituon.petpet.bot.qq.permission.ContactPermission;
import moe.dituon.petpet.bot.utils.Cooler;
import moe.dituon.petpet.core.context.RequestContext;
import moe.dituon.petpet.core.element.PetpetModel;
import moe.dituon.petpet.core.element.PetpetTemplateModel;
import moe.dituon.petpet.core.imgres.ImageResourceMap;
import moe.dituon.petpet.script.PetpetJsScriptModel;
import moe.dituon.petpet.script.PetpetScriptModel;
import net.mamoe.mirai.contact.AnonymousMember;
import net.mamoe.mirai.contact.Member;
import net.mamoe.mirai.event.events.GroupMessageEvent;
import net.mamoe.mirai.message.data.*;
import org.jetbrains.annotations.Nullable;

import java.net.MalformedURLException;
import java.util.*;

@Slf4j
public class GroupMessageEventHandler extends MiraiMessageEventHandler {
    protected final String command;
    private final MiraiBotService service;
    private static final String[] EMPTY_STRING_ARRAY = new String[0];
    protected final boolean respondNudgeWhenCooldown;

    public GroupMessageEventHandler(MiraiBotService service) {
        this.service = service;
        this.command = service.miraiConfig.getCommand();
        this.respondNudgeWhenCooldown = "[nudge]".equals(service.miraiConfig.getInCoolDownMessage());
    }

    public void handle(GroupMessageEvent event) {
        var handler = new Handler(event);
        handler.handle();
    }

    protected class Handler {
        protected final GroupMessageEvent event;
        protected final String message;
        protected final ContactPermission permission;
        protected PetpetModel templateModel;
        protected boolean resultDefault = false;

        public Handler(GroupMessageEvent event) {
            this.event = event;
            this.permission = service.getPermission(event.getGroup().getId());

            var messageBuilder = new StringBuilder();
            for (SingleMessage ele : event.getMessage()) {
                if (ele instanceof PlainText) {
                    var text = (PlainText) ele;
                    messageBuilder.append(text.getContent()).append(' ');
                } else if (ele instanceof Image) {
                    int[] sourceIds = event.getMessage().get(MessageSource.Key).getIds();
                    if (sourceIds.length == 0) {
                        continue;
                    }

                    // message id = group id + source id
                    long id = event.getGroup().getId() + sourceIds[0];
                    service.getImageCachePool().put(id, Image.queryUrl((Image) ele));
                }
            }
            this.message = messageBuilder.toString();
        }

        public void handle() {
            if (handleOpCommand()) return;
            handleCommand();
        }

        protected boolean handleOpCommand() {
            if (!hasGroupPermission(this.event.getSender())) return false;

            if (!this.message.startsWith(command)) {
                return false;
            }

            var commandOperation = this.message.substring(command.length());
            String result;
            try {
                result = this.permission.handleEditCommand(commandOperation);
                if (result == null) {
                    return false;
                }
            } catch (IllegalArgumentException | IllegalStateException e) {
                result = e.getMessage();
            }
            this.sendMessage(command + ' ' + result);

            return true;
        }

        protected void handleCommand() {
            var commandHead = service.miraiConfig.getCommandHead();
            if (message.startsWith(command)) { // /pet (id?) (param?)
                if ((permission.getCommandPermission() & ContactPermission.COMMAND) == 0) {
                    // no permission
                    return;
                }
                var idAndParam = message.substring(command.length()).trim();
                if (idAndParam.isBlank()) { // /pet
                    boolean isUseDefaultTemplate = service.miraiConfig.getDefaultReplyType() != ReplyType.RANDOM;
                    handleIdAndParam(idAndParam, false, isUseDefaultTemplate);
                    return;
                }
                handleIdAndParam(idAndParam, false, false); // (id?) (param?)
            } else if (commandHead.isEmpty() || message.startsWith(commandHead)) { // #(id?) (param?)
                if ((permission.getCommandPermission() & ContactPermission.COMMAND_HEAD) == 0) {
                    // no permission
                    return;
                }
                var idAndParam = message.substring(commandHead.length()).trim();
                handleIdAndParam(idAndParam, true, false);
            }
        }

        /**
         * @param idAndParam   (id?) (param?)
         * @param idIsRequired if id is undefined, use random id or default template
         */
        protected void handleIdAndParam(String idAndParam, boolean idIsRequired, boolean useDefaultTemplate) {
            var tokens = idAndParam.split(" +"); // [id, ...params]
            String idOrAlias = tokens.length > 0 ? tokens[0].replace("\n", "") : "";
            var model = resolveTemplateModel(idOrAlias, idIsRequired, useDefaultTemplate);

            if (model == null) return;

            if (Cooler.isLocked(event.getGroup().getId()) || Cooler.isLocked(event.getSender().getId())) {
                if (respondNudgeWhenCooldown && !(event.getSender() instanceof AnonymousMember)) {
                    event.getSender().nudge().sendTo(event.getGroup());
                    return;
                }
                if (service.miraiConfig.getInCoolDownMessage().isBlank()) return;
                sendMessage(service.miraiConfig.getInCoolDownMessage());
                return;
            }
            this.templateModel = model;

            // TODO: cache request map
            var context = buildRequestContext(
                    tokens.length > 0 ? Arrays.copyOfRange(tokens, 1, tokens.length) : EMPTY_STRING_ARRAY,
                    tokens.length > 0 ? idAndParam.substring(idOrAlias.length()).trim() : ""
            );
            sendImage(model, context);

            if (!resultDefault) {
                Cooler.lock(event.getGroup().getId(), permission.getCooldownTime());
                Cooler.lock(event.getSender().getId(), service.miraiConfig.getUserCoolDown());
            }
        }

        /**
         * 根据 ID 或别名解析模板模型，如果找不到则根据参数选择默认模板或随机模板。
         */
        private @Nullable PetpetModel resolveTemplateModel(String idOrAlias, boolean idIsRequired, boolean useDefaultTemplate) {
            if (idOrAlias.isEmpty()) {
                if (idIsRequired) return null;
                this.resultDefault = useDefaultTemplate;
                return useDefaultTemplate ? service.getDefaultTemplate() : service.randomTemplate();
            }

            var model = service.getTemplate(idOrAlias);
            if (model == null && !idIsRequired) {
                this.resultDefault = useDefaultTemplate;
                model = useDefaultTemplate ? service.getDefaultTemplate() : service.randomTemplate();
            }
            return model;
        }

        /**
         * @param params [id, ...params]
         */
        protected RequestContext buildRequestContext(String[] params, String rawParams) {
            var textMap = new HashMap<String, String>(params.length + 8);
            textMap.put("raw", rawParams);
            for (int i = 0; i < params.length; i++) {
                // text variable e.g. text${1}
                textMap.put(String.valueOf(i), params[i]);
            }

            boolean ignoreAt = false;
            // default subject is BOT
            Member from = event.getGroup().getBotAsMember();
            // default target is sender
            Member to = event.getSender();
            var customImageList = new ArrayList<String>(event.getMessage().size() - 1);
            for (SingleMessage ele : event.getMessage()) {
                if (ele instanceof At && !ignoreAt) {
                    if ((permission.getCommandPermission() & ContactPermission.AT) == 0) {
                        continue;
                    }
                    At at = (At) ele;
                    // ignored at self
                    if (at.getTarget() == event.getSender().getId()) continue;

                    // form = sender; to = at target
                    from = event.getSender();
                    to = event.getGroup().get(at.getTarget());
                    assert to != null;
                } else if (ele instanceof QuoteReply) {
                    if ((permission.getCommandPermission() & ContactPermission.IMAGE) == 0) {
                        continue;
                    }
                    var reply = (QuoteReply) ele;
                    var replySourceIds = reply.getSource().getIds();
                    if (replySourceIds.length == 0) continue;
                    // message id = group id + source id
                    long msgId = event.getGroup().getId() + replySourceIds[0];

                    var sourceImgSrc = service.getImageCachePool().get(msgId);
                    if (sourceImgSrc == null) continue;
                    customImageList.add(sourceImgSrc);
                    from = event.getSender();
                    ignoreAt = true;
                } else if (ele instanceof Image) {
                    if ((permission.getCommandPermission() & ContactPermission.IMAGE) == 0) {
                        continue;
                    }
                    customImageList.add(Image.queryUrl((Image) ele));
                    from = event.getSender();
                    ignoreAt = true;
                }
            }

            String toName = getNameOrNick(to);
            // get template element expected size to query avatar image
            String toUrl = getAvatarUrl(to.getId(), TO_KEY);
            String fromName = getNameOrNick(from);
            String fromUrl = getAvatarUrl(to.getId(), FROM_KEY);
            List<String> randomImageList = null;

            // customImageList: [to, form, ...random]
            if (!customImageList.isEmpty()) {
                if (customImageList.size() > 1) {
                    // stay from sender name
                    fromUrl = customImageList.get(1);
                    if (customImageList.size() > 2) {
                        // TODO: RandomImageResource
//                        randomImageList = customImageList.subList(2, customImageList.size());
                    }
                }
                toUrl = customImageList.get(0);
                toName = "这个";
            }

            textMap.put(FROM_KEY, fromName);
            textMap.put(TO_KEY, toName);
            textMap.put(GROUP_KEY, event.getGroup().getName());
            textMap.put(BOT_KEY, getNameOrNick(event.getGroup().getBotAsMember()));
            textMap.put(FROM_ID_KEY, String.valueOf(from.getId()));
            textMap.put(TO_ID_KEY, String.valueOf(to.getId()));
            textMap.put(GROUP_ID_KEY, String.valueOf(event.getGroup().getId()));
            textMap.put(BOT_ID_KEY, String.valueOf(event.getBot().getId()));

            try {
                return new RequestContext(
                        new ImageResourceMap(Map.of(
                                FROM_KEY, QQAvatarRequester.getAvatarResource(fromUrl),
                                TO_KEY, QQAvatarRequester.getAvatarResource(toUrl),
                                GROUP_KEY, QQAvatarRequester.getAvatarResource(
                                        getAvatarUrl(event.getGroup().getId(), GROUP_KEY)
                                ),
                                BOT_KEY, QQAvatarRequester.getAvatarResource(
                                        getAvatarUrl(event.getBot().getId(), BOT_KEY)
                                )
                        )),
                        textMap
                );
            } catch (MalformedURLException e) {
                // never
                throw new IllegalStateException(e);
            }
        }

        protected String getAvatarUrl(long qqId, String key) {
            return QQAvatarRequester.getAvatarUrlString(
                    qqId,
                    service.getTemplateExpectedSize(templateModel)
                            .getOrDefault(key, 640)
            );
        }

        protected void sendImage(PetpetModel model, RequestContext context) {
            if (model instanceof PetpetScriptModel) {
                var scriptModel = (PetpetScriptModel) model;
                var basePath = scriptModel instanceof PetpetJsScriptModel ?
                        ((PetpetJsScriptModel) scriptModel).basePath : null;
                var scriptEvent = new ScriptMiraiBotSendEvent(this.event, context, basePath);
                if (scriptModel.getEventManager().has("bot_send")) {
                    scriptModel.getEventManager().trigger("bot_send", scriptEvent);
                }
                this.event.getGroup().sendMessage(new QuoteReply(event.getMessage())
                        .plus(scriptEvent.getResponseMessage())
                );
            } else if (model instanceof PetpetTemplateModel) {
                var img = model.draw(context);
                try {
                    sendMessage(
                            service.uploadImage(img.bytes, event.getGroup())
                    );
                } catch (Exception ex) {
                    sendMessage("上传图像时出错: " + ex.getMessage());
                    log.warn("上传图像时出错", ex);
                }
            } else {
                // never
                throw new IllegalStateException();
            }
        }

        protected void sendMessage(String msg) {
            event.getGroup().sendMessage(new QuoteReply(event.getMessage()).plus(msg));
        }

        protected void sendMessage(Image image) {
            event.getGroup().sendMessage(new QuoteReply(event.getMessage()).plus(image));
        }
    }
}

File: bot/onebot/src/main/kotlin/moe/dituon/petpet/bot/qq/onebot/GroupMessageEventHandler.kt
package moe.dituon.petpet.bot.qq.onebot

import cn.evolvefield.onebot.sdk.entity.GroupSender
import cn.evolvefield.onebot.sdk.event.message.GroupMessageEvent
import cn.evolvefield.onebot.sdk.util.CQCode
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import lombok.extern.slf4j.Slf4j
import moe.dituon.petpet.bot.MessageEventHandler
import moe.dituon.petpet.bot.qq.ReplyType
import moe.dituon.petpet.bot.qq.avatar.QQAvatarRequester
import moe.dituon.petpet.bot.qq.permission.ContactPermission
import moe.dituon.petpet.bot.utils.Cooler
import moe.dituon.petpet.core.context.RequestContext
import moe.dituon.petpet.core.element.PetpetModel
import moe.dituon.petpet.core.element.PetpetTemplateModel
import moe.dituon.petpet.core.imgres.ImageResource
import moe.dituon.petpet.core.imgres.ImageResourceMap
import moe.dituon.petpet.core.utils.image.EncodedImage
import moe.dituon.petpet.script.PetpetJsScriptModel
import moe.dituon.petpet.script.PetpetScriptModel
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.net.MalformedURLException
import java.util.*
import kotlin.collections.set
import kotlin.streams.toList


@Slf4j
open class GroupMessageEventHandler(private val service: OnebotBotService): MessageEventHandler() {
    companion object {
        private val log: Logger = LoggerFactory.getLogger(this::class.java)
    }

    protected val command: String = service.getOnebotConfig().command
    protected val respondNudgeWhenCooldown: Boolean = "[nudge]" == service.onebotConfig.inCoolDownMessage

    suspend fun handle(event: GroupMessageEvent) {
        val handler = Handler(event)
        handler.handle()
    }

    open inner class Handler(
        protected val event: GroupMessageEvent
    ) {
        protected lateinit var message: String
        protected val permission: ContactPermission = service.getPermission(event.groupId)
        protected var templateModel: PetpetModel? = null
        protected var resultDefault: Boolean = false

        init {
            init()
        }

        protected fun init() {
            if (event.sender == null) {
                message = ""
                return
            }
            val messageBuilder = StringBuilder()
            val messages = if (event.isJsonMessage) {
                event.json.asJsonObject.getAsJsonArray("message")
            } else {
                CQCode.toJson(event.message)
            }
            for (ele in messages) {
                val msg = ele.asJsonObject
                val type = msg["type"].asString

                when (type) {
                    "text" -> messageBuilder.append(
                        msg["data"].asJsonObject["text"].asString
                    ).append(' ')

                    "image" -> {
                        // id = group id + message id
                        val id = event.groupId + event.messageId
                        service.imageCachePool[id] = msg["data"].asJsonObject["url"].asString
                    }
                }
            }
            this.message = messageBuilder.toString()
        }

        suspend fun handle() {
            if (message.isBlank() || event.sender!!.userId == globalBotInstance.id.toString()) return
            if (handleOpCommand()) return
            handleCommand()
        }

        protected fun hasGroupPermission(user: GroupSender): Boolean {
            return user.role != "member"
        }

        protected suspend fun handleOpCommand(): Boolean {
            if (!hasGroupPermission(event.sender!!)) return false

            if (!message.startsWith(command)) {
                return false
            }

            val commandOperation = message.substring(command.length)
            var result: String?
            try {
                result = permission.handleEditCommand(commandOperation)
                if (result == null) {
                    return false
                }
            } catch (e: IllegalArgumentException) {
                result = e.message
            } catch (e: IllegalStateException) {
                result = e.message
            }
            this.sendMessage("$command $result")

            return true
        }

        protected suspend fun handleCommand() {
            val commandHead = service.onebotConfig.commandHead
            if (message.startsWith(command)) { // /pet (id?) (param?)
                if ((permission.commandPermission and ContactPermission.COMMAND) == 0) {
                    // no permission
                    return
                }
                val idAndParam = message.substring(command.length).trim { it <= ' ' }
                if (idAndParam.isBlank()) { // /pet
                    val isUseDefaultTemplate = service.config.defaultReplyType != ReplyType.RANDOM
                    handleIdAndParam(idAndParam, false, isUseDefaultTemplate)
                    return
                }
                handleIdAndParam(idAndParam, false, false) // (id?) (param?)
            } else if (commandHead.isEmpty() || message.startsWith(commandHead)) { // #(id?) (param?)
                if ((permission.commandPermission and ContactPermission.COMMAND_HEAD) == 0) {
                    // no permission
                    return
                }
                val idAndParam: String = message.substring(commandHead.length).trim { it <= ' ' }
                handleIdAndParam(idAndParam, true, false)
            }
        }

        /**
         * @param idAndParam   (id?) (param?)
         * @param idIsRequired if id is undefined, use random id or default template
         */
        protected suspend fun handleIdAndParam(idAndParam: String, idIsRequired: Boolean, useDefaultTemplate: Boolean) {
            val tokens = idAndParam  // [id, ...params]
                .split(" +".toRegex())
                .dropLastWhile { it.isEmpty() }
                .toTypedArray()
            val idOrAlias = if (tokens.isNotEmpty()) tokens[0].replace("\n", "") else ""
            val model = resolveTemplateModel(idOrAlias, idIsRequired, useDefaultTemplate) ?: return

            if (Cooler.isLocked(event.groupId) || Cooler.isLocked(event.sender!!.userId)) {
                if (respondNudgeWhenCooldown) {
                    // nudge
                    sendJsonMessage("[{\"type\":\"touch\",\"data\":{\"id\":${event.sender!!.userId}}}]")
                    return
                }
                if (service.onebotConfig.inCoolDownMessage.isBlank()) return
                sendMessage(service.onebotConfig.inCoolDownMessage)
                return
            }

            this.templateModel = model

            // TODO: cache request map
            val context = buildRequestContext(
                if (tokens.isNotEmpty()) Arrays.copyOfRange(tokens, 1, tokens.size) else emptyArray(),
                if (tokens.isNotEmpty()) idAndParam.substring(idOrAlias.length).trim { it <= ' ' } else ""
            )
            sendImage(model, context)

            if (!resultDefault) {
                Cooler.lock(event.groupId, permission.cooldownTime)
                Cooler.lock(event.sender!!.userId, service.onebotConfig.userCooldownTime)
            }
        }

        /**
         * 根据 ID 或别名解析模板模型，如果找不到则根据参数选择默认模板或随机模板。
         */
        private fun resolveTemplateModel(
            idOrAlias: String,
            idIsRequired: Boolean,
            useDefaultTemplate: Boolean
        ): PetpetModel? {
            if (idOrAlias.isEmpty()) {
                if (idIsRequired) return null
                return if (useDefaultTemplate) {
                    this.resultDefault = true
                    service.defaultTemplate
                } else service.randomTemplate()
            }

            var model = service.getTemplate(idOrAlias)
            if (model == null && !idIsRequired) {
                model = if (useDefaultTemplate) {
                    this.resultDefault = true
                    service.defaultTemplate
                } else service.randomTemplate()
            }
            return model
        }

        fun getNameOrNick(sender: GroupSender) = sender.card.ifBlank {
            sender.nickname
        }

        /**
         * @param params [id, ...params]
         */
        protected fun buildRequestContext(params: Array<String>, rawParams: String): RequestContext {
            val textMap = HashMap<String, String>(params.size + 8)
            textMap["raw"] = rawParams
            for (i in params.indices) {
                // text variable e.g. text${1}
                textMap[(i).toString()] = params[i]
            }

            var ignoreAt = false
            // default subject is BOT
            var fromName: String? = null
            var fromUrl: String? = null
            var fromId: String? = null
            // default target is sender
            var toName: String? = null
            var toUrl: String? = null
            var toId: String? = null
            val messages = event.json.asJsonObject["message"].asJsonArray
            val customImageList: ArrayList<String> = ArrayList<String>(messages.size())
            for (ele in messages) {
                val msg = ele.asJsonObject
                val type = msg["type"].asString
                when (type) {
                    "at" -> {
                        if (ignoreAt || (permission.commandPermission and ContactPermission.AT) == 0) {
                            continue
                        }

                        val target = msg["data"].asJsonObject["qq"].asString
                        // ignored at all or self
                        if (target == "all" || target == event.sender!!.userId) {
                            continue
                        }

                        // form = sender; to = at target
                        fromName = getNameOrNick(event.sender!!)
                        fromId = event.sender!!.userId
                        fromUrl = getAvatarUrl(fromId, "from")
                        toName = target //TODO
                        toUrl = getAvatarUrl(target, "to")
                        toId = target
                    }

                    "reply" -> {
                        if ((permission.commandPermission and ContactPermission.IMAGE) == 0) {
                            continue
                        }

                        val id = msg["data"].asJsonObject["id"].asString.toLong()
                        val sourceImgSrc = service.imageCachePool[event.groupId + id] ?: continue
                        customImageList.add(sourceImgSrc)

                        fromName = getNameOrNick(event.sender!!)
                        fromUrl = getAvatarUrl(event.sender!!.userId, "from")
                        fromId = event.sender!!.userId
                        ignoreAt = true
                    }

                    "image" -> {
                        if ((permission.commandPermission and ContactPermission.IMAGE) == 0) {
                            continue
                        }
                        customImageList.add(msg["data"].asJsonObject["url"].asString)

                        fromName = getNameOrNick(event.sender!!)
                        fromId = event.sender!!.userId
                        fromUrl = getAvatarUrl(fromId, "from")
                        ignoreAt = true
                    }
                }
            }

            // customImageList: [to, form, ...random]
            if (customImageList.isNotEmpty()) {
                if (customImageList.size > 1) {
                    // stay from sender name
                    fromUrl = customImageList[1]
                    fromId = fromUrl
                    if (customImageList.size > 2) {
                        // TODO: RandomImageResource
//                        randomImageList = customImageList.subList(2, customImageList.size());
                    }
                }
                toUrl = customImageList[0]
                toId = toUrl
                toName = "这个"
            }

            fromName = fromName ?: "bot" //TODO
            fromUrl = fromUrl ?: getAvatarUrl(globalBotInstance.id.toString(), "from")
            fromId = fromId ?: event.sender!!.userId
            toName = toName ?: event.sender!!.card
            toUrl = toUrl ?: getAvatarUrl(event.sender!!.userId, "to")
            toId = toId ?: event.sender!!.userId

            // TODO: 按需获取数据
            textMap[FROM_KEY] = fromName
            textMap[TO_KEY] = toName
            textMap[GROUP_KEY] = event.groupId.toString() //TODO
            textMap[BOT_KEY] = "bot" //TODO
            textMap[FROM_ID_KEY] = fromId
            textMap[TO_ID_KEY] = toId
            textMap[GROUP_ID_KEY] = event.groupId.toString()
            textMap[BOT_ID_KEY] = globalBotInstance.id.toString()

            try {
                return RequestContext(
                    ImageResourceMap(
                        mapOf<String, ImageResource>(
                            FROM_KEY to QQAvatarRequester.getAvatarResource(fromUrl),
                            TO_KEY to QQAvatarRequester.getAvatarResource(toUrl),
                            GROUP_KEY to QQAvatarRequester.getAvatarResource(
                                getAvatarUrl(event.groupId, GROUP_KEY)
                            ),
                            BOT_KEY to QQAvatarRequester.getAvatarResource(
                                getAvatarUrl(globalBotInstance.id, BOT_KEY)
                            )
                        )
                    ),
                    textMap
                )
            } catch (e: MalformedURLException) {
                // never
                throw IllegalStateException(e)
            }
        }

        protected fun getAvatarUrl(qqId: Long, key: String?) = getAvatarUrl(qqId.toString(), key)
        protected fun getAvatarUrl(qqId: String, key: String?): String {
            return QQAvatarRequester.getAvatarUrlString(
                qqId,
                service.getTemplateExpectedSize(templateModel)
                    .getOrDefault(key, 640)
            )
        }

        protected suspend fun sendImage(model: PetpetModel?, context: RequestContext?) {
            if (model is PetpetScriptModel) {
                val basePath = if (model is PetpetJsScriptModel) model.basePath else null
                val scriptEvent = ScriptOnebotSendEvent(service, this.event, context, basePath)
                if (model.eventManager.has("bot_send")) {
                    model.eventManager.trigger("bot_send", scriptEvent)
                }
                val msgList = scriptEvent.responseMessage ?: return
                if (scriptEvent.isResponseInForward) {
                    globalBotInstance.sendGroupForwardMsg(event.groupId, "[" + msgList.stream().map {
                        """
                            {
                                "type": "node",
                                "data": {
                                    "user_id": "${globalBotInstance.id}",
                                    "nickname": "Petpet!",
                                    "content": $it
                                }
                            }
                            """.trimIndent()
                    }.toList().joinToString(",") + "]")
                } else {
                    for (msgGroup in msgList) {
                        sendJsonMessage(msgGroup)
                    }
                }
            } else if (model is PetpetTemplateModel) {
                val img = model.draw(context)
                try {
                    sendMessage(img)
                } catch (ex: Exception) {
                    sendMessage("上传图像时出错: " + ex.message)
                    log.warn("上传图像时出错", ex)
                }
            }
        }

        protected suspend fun sendJsonMessage(json: String) {
            globalBotInstance.sendGroupMsg(event.groupId, json, false)
        }

        protected suspend fun sendMessage(msg: String) {
            globalBotInstance.sendGroupMsg(
                event.groupId, """[
                {"type": "reply","data": {"id": "${event.messageId}"}},
                {"type": "text","data": {"text": ${Json.encodeToString(String.serializer(), msg)}}}
            ]""".trimIndent(), false
            )
        }

        protected suspend fun sendMessage(image: EncodedImage) {
            globalBotInstance.sendGroupMsg(
                event.groupId, """[
                {"type": "reply","data": {"id": "${event.messageId}"}},
                {"type": "image","data": {"file": "${service.putImage(image)}"}}
                ]""".trimIndent(), false
            )
        }
    }
}

